### PR TITLE
[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103784
* #103783
* #103782
* #103781
* __->__ #103609

Users may prefix the keys optim state_dict. Using`optim.state_dict()` to get the initial states is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

### CODE REVIEW

File: torch/distributed/fsdp/_optim_utils.py [hunk 1]
```diff
@@ -475,12 +458,21 @@ def _flatten_optim_state_dict(
...
-                assert len(fqns) == 1
-                local_wrapped_fqn = local_state_clean_fqns.get(fqn, "")
-                if local_wrapped_fqn:
-                    flat_osd_state[key] = copy.deepcopy(
-                        local_state_dict[local_wrapped_fqn]
+                assert len(fqns) == 1 and use_orig_params
+                flat_osd_state[key] = copy.deepcopy(
+                    tree_map_only(
+                        torch.Tensor,
+                        lambda t: t.cpu()
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:465]: nit: Shall we preserve the comment that we are specially keeping dim-0 states like `"step"` on CPU? (same with above in `_broadcast_processed_state()`)

File: torch/distributed/fsdp/_optim_utils.py [hunk 2]
```diff
@@ -475,12 +458,21 @@ def _flatten_optim_state_dict(
...
+                        if t.dim() == 0
+                        else torch.zeros(0, device=t.device, dtype=t.dtype),
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:467]: Is there any concern that there may be a mixture of CPU and GPU states? (I am guessing that `t.device` might be GPU?)

File: torch/distributed/fsdp/_optim_utils.py [hunk 3]
```diff
@@ -475,12 +458,21 @@ def _flatten_optim_state_dict(
             if flat_state:
                 flat_osd_state[key] = flat_state
             elif optim is not None:  # NamedOptimizer or KeyedOptimizer case.
-                assert len(fqns) == 1
-                local_wrapped_fqn = local_state_clean_fqns.get(fqn, "")
-                if local_wrapped_fqn:
-                    flat_osd_state[key] = copy.deepcopy(
-                        local_state_dict[local_wrapped_fqn]
+                assert len(fqns) == 1 and use_orig_params
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:461]: nit: Should we add an assert message here just in case? For example, it may be helpful to at least include `"{len(fqns)} {fqns}` (so we know whether it was `len(fqns) !=` or `not use_orig_params` and have some sense of the FQNs).

**Review by awgu — APPROVED** [review]
Thanks for the fix!

File: torch/distributed/fsdp/_optim_utils.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:465]: I don't think this is an issue as the step is generally on CPU and other states may be moved to GPU. Moreover, this one is specific to the zero-size states which are not going to be used because the parameters are empty.

File: torch/distributed/fsdp/_optim_utils.py (hunk 2 above)
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py]: I don't think this is an issue as the step is generally on CPU and other states may be moved to GPU. Moreover, this one is specific to the zero-size states which are not going to be used because the parameters are empty.

**Comment by fegin** [comment]
The failing test is not related.
```
Usage:   
  pip install [options] <requirement specifier> [package-index-options] ...
  pip install [options] -r <requirements file> [package-index-options] ...
  pip install [options] [-e] <vcs project url> ...
  pip install [options] [-e] <local project path> ...
  pip install [options] <archive url/path> ...

no such option: --install-option
Error: Process completed with exit code 2.
```
